README.md
<h1 align="center">KudaOpenAPI Integration in Flutter</h1>

<!-- Status -->

<h4 align="center">
	KudaOpenAPI for flutter for seemless banking via Kudaopenapi.
</h4>

<hr>

<p align="center">
  <a href="#dart-getting-started">Getting Started</a> &#xa0; | &#xa0;
  <a href="#dart-installation">Installation</a> &#xa0; | &#xa0;
  <a href="#sparkles-usage">Usage</a> &#xa0; | &#xa0;
  <a href="#memo-license">License</a> &#xa0; | &#xa0;
  <a href="https://github.com/giftbalogun" target="_blank">Author</a>
</p>

<br>

## :dart: Getting Started

Enable your product for local transactions with the KudaOpenAPI! With the KUDA Open APIs you can embed services unto your platform and connect your customers to a wide range of banking services.

Before you proceed, ensure you have a [Kuda Business account](https://business.kuda.com/)!. You can link this account to your profile to get approved for live. 

Get your token from the [Developer Dashboard](https://developer.kuda.com/)!.

## :dart: Installation

To use this plugin, add `kudaopenapi` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

Then initialize the plugin preferably in the `initState` of your widget.

``` dart
import 'package:kudaopenapi/kudaopenapi.dart';

class _MyHomePageState extends State<MyHomePage> {
	var baseurl = 'TEST_OR_LIVE_URL';
	var email = 'EMAIL';
	var apikey = 'APIKEY';
	@override
	void initState() {
		ApiService.initialize(baseurl, email, apikey);
		super.initState();
	}
}
```

No other configuration required&mdash;the plugin works out of the box.

## :sparkles: Usage

Make Request to the API using this

``` dart
String trackingReference = Random().nextInt(100000).toString();

Map<String, dynamic> data = {
  'ClientAccountNumber': "00000000000",
  'beneficiarybankCode': "000014",
  'beneficiaryAccount': "000000000000",
  'beneficiaryName': "GIFT IWOKURA BALOGUN",
  'amount': "5000",
  'narration': "Payment Flutter",
  'nameEnquirySessionID': "00000000000000000",
  'trackingReference': Random().nextInt(100000).toString(),
  'senderName': "Gift Balogun",
}; //Format for sending data through

String requestRef = Random().nextInt(100000).toString();
```

To get the list of banks with the kudaopenapi

``` dart
import 'package:kudaopenapi/kudaopenapi.dart';

@override
  Widget build(BuildContext context) {
  return MaterialApp(
    home: Scaffold(
        appBar: AppBar(title: const Text("Kuda Get Bank API")),
        body: FutureBuilder<BankResponse>(
          future: KudaBank().getBankList(requestRef),
          builder: (context, snapshot) {
            if (snapshot.hasData) {
              return ListView.builder(
                itemCount: snapshot.data!.data!.banks.length,
                itemBuilder: (context, index) {
                  return ListTile(
                    leading: const Icon(Icons.list),
                    trailing: Text(
                      snapshot.data!.data!.banks[index].bankCode.toString(),
                      style: TextStyle(color: Colors.green, fontSize: 15),
                    ),
                    title: Text(
                      snapshot.data!.data!.banks[index].bankName.toString(),
                    ),
                  );
                },
              );
            } else if (snapshot.hasError) {
              return Text(snapshot.error.toString());
            }
            return CircularProgressIndicator();
          },
        )
      ),
  );
}
```

To get the details of an account

``` dart
import 'package:kudaopenapi/kudaopenapi.dart';

class _MyHomePageState extends State<MyHomePage> {
  var baseurl = '';
  var email = '';
  var apikey = '';

  @override
  void initState() {
    ApiService.initialize(baseurl, email, apikey);
    super.initState();
  }

  String trackingReference = Random().nextInt(100000).toString();

  // Set the request data
  Map<String, dynamic> data = {
    'beneficiaryAccountNumber': '1413800836',
    'beneficiaryBankCode': '000014',
    'SenderTrackingReference': '',
    'isRequestFromVirtualAccount': true,
  };

  String requestRef = Random().nextInt(100000).toString();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Kuda Name Inquiry API")),
      body: Center(
        child: FutureBuilder(
          future: KudaBank().name_inquiry(data, requestRef),
          builder: (BuildContext context, AsyncSnapshot snapshot) {
            if (snapshot.connectionState == ConnectionState.done) {
              if (snapshot.hasError) {
                return Text('Error: ${snapshot.error}');
              } else {
                var something = snapshot.data.data!;
                return Text('sessionID: ${something.nameInquiry!.sessionID}');
              }
            } else {
              return CircularProgressIndicator();
            }
          },
        ),
      ),
    );
  }
}
```

To get retrieve virtual account details with the kudaopenapi

``` dart
import 'package:kudaopenapi/kudaopenapi.dart';

@override
Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(title: const Text("KudaOpenApi-Retrieve")),
        body: Center(
          child: Column(mainAxisAlignment: MainAxisAlignment.spaceEvenly, children: [
            FutureBuilder(
              future: KudaBank().retrieve_virtual_account(data, requestRef),
              builder: (BuildContext context, AsyncSnapshot snapshot) {
                if (snapshot.connectionState == ConnectionState.done) {
                  if (snapshot.hasError) {
                    return Text('Error: ${snapshot.error}');
                  } else {
                    var something = snapshot.data.data!;
                    return Column(
                      children: [
                        Text('AccountNumber: ${something!.rvirtualAccount!.accountnumber!}'),
                        Text('Email.: ${something!.rvirtualAccount!.email}'),
                        Text('Phone Number.: ${something!.rvirtualAccount!.phonenumber}'),
                        Text('AccountName.: ${something!.rvirtualAccount!.accountname}'),
                        Text('Fullname.: ${something!.rvirtualAccount!.firstname}'),
                        Text('Tracking Ref.: ${something!.rvirtualAccount!.trackingref}'),
                      ],
                    );
                  }
                } else {
                  return CircularProgressIndicator();
                }
              },
            ),
          ]
        ),
      ),
    );
  }
}
```

To get list of GiftCards Offered by Kuda

```dart
Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Gift Cards')),
      body: FutureBuilder<GiftCardResponse>(
        future: KudaGiftCard().list_gift_cards(requestRef),
        builder: (context, snapshot) {
          if (snapshot.connectionState == ConnectionState.waiting) {
            return Center(child: CircularProgressIndicator());
          } else if (snapshot.hasError) {
            return Center(child: Text('Error: ${snapshot.error}'));
            
          } else if (!snapshot.hasData || snapshot.data!.data.giftItems.isEmpty) {
            return Center(child: Text('No gift cards available'));
          } else {
            return ListView.builder(
              itemCount: snapshot.data!.data.giftItems.length,
              itemBuilder: (context, index) {
                final giftCard = snapshot.data!.data.giftItems[index];
                return ListTile(
                  leading: Image.network(giftCard.image),
                  title: Text(giftCard.package),
                  subtitle: Text('Category: ${giftCard.category}'),
                  trailing: Text(
                    'Country: ${giftCard.countries.length}',
                    style: TextStyle(color: Colors.red),
                  ),
                  onTap: () {
                    // Handle gift card item tap
                  },
                );
              },
            );
          }
        },
      ),
    );
  }
```

## :memo: License

This project is under license from MIT. For more details, see the [LICENSE](LICENSE) file.

## Social Presense
Follow me on social media
[Medium](https://medium.com/@giftbalogun)!
[Twitter](https://twitter.com/am_de_one)!
[Instagram](https://www.instagram.com/am_thd_one/)!
[LinkedIn](https://www.linkedin.com/in/gift-balogun-907103160/)!
[Porfolio](https://giftbalogun.name.ng/)!

&#xa0;

<a href="#top">Back to top</a>

CHANGELOG.md
## 0.0.1

-Initial Package Realease (Beta)

## 0.0.2

-Package Correction (Beta)

## 0.0.3

-Updated Kuda Service with Response (Beta)
-Added Model to Services

## 0.0.4

-Updated Kuda Service with Response (Beta)
-Added More Model to Services
-Added Transactions History Log
-Updated the pubspec.yaml
-Updated README.md
-Updated Service Type

## 0.0.5

-Updated Service typed
-Added Kuda Savings API
-Updated kuda_bank.dart
-Updated README.md
-Updated Models

## 0.0.6

-Updated Service typed
-Added Kuda Savings API
-Updated kuda_bank.dart
-Updated README.md
-Updated Models
-Added Documentations
-Updated kudaopenapi.dart
-added kuda savings

## 0.0.7

-Updated Service Type
-Added Kuda_Card.dart
-Updated Model
-Updated Documentation
-Code Fixes
-Added Kuda Gift Cards Service
-Added Kuda Gift Cards Model
-Updated all libraries

## 0.0.8

-Updated README.md
-Updated all library files
-Code Fixes
-Updated API Calls

## 0.0.9

-Updated Example with 2 new details
-Updated Model Library
-Updated API library
-Updated Services Library
-Bug fixes and improment

## 0.1.0

-Updated service type library
-update kuda_bank services
-updated example program
-updated all library
-chnaged model to response
-updated UI for example.

## 0.1.1

-no major change

## 0.1.2

-bug fixes and improvement

## 0.1.3

-Updated Billing Library

## 0.1.4

-Bug fixes and improvemnt

## 0.1.5

-Bug fixes and improvemnt

## 0.1.6

-Bug fixes and improvemnt
-added library

## 0.1.7

-Bug fixes and improvemnt
-added API Response


example/lib/second.dart
// ignore_for_file: non_constant_identifier_names, library_private_types_in_public_api, use_key_in_widget_constructors

import 'dart:math';

import 'package:flutter/material.dart';
import 'package:kudaopenapi/kudaopenapi.dart';

class TransferPage extends StatefulWidget {
  @override
  _TransferPageState createState() => _TransferPageState();
}

class _TransferPageState extends State<TransferPage> {
  final GlobalKey<FormState> _key = GlobalKey<FormState>();
  late TextEditingController _accountController;
  late TextEditingController _amountController;
  final TextEditingController _narrationController = TextEditingController();

  String? selectedBankCode;
  List<Bank> bankList = [];
  String? accountName;
  String? session_id;
  String? bank_codes;
  String? response_codes;
  String requestRef = Random().nextInt(100000).toString();

  @override
  void initState() {
    super.initState();
    fetchBankList();
    _accountController = TextEditingController();
    _amountController = TextEditingController();
  }

  @override
  void dispose() {
    _accountController.dispose();
    _amountController.dispose();
    _narrationController.dispose();
    super.dispose();
  }

  Future<void> fetchBankList() async {
    try {
      BankResponse bankResponse = await KudaBank().getBankList(requestRef);
      setState(() {
        bankList = bankResponse.data.banks;
        selectedBankCode =
            bankList.isNotEmpty ? bankList[0].bankCode.toString() : null;
      });
    } catch (error) {
      print('Error fetching bank list: $error');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Account Transfer'),
      ),
      body: SafeArea(
        minimum: const EdgeInsets.all(16),
        child: _buildTransferForm(),
      ),
    );
  }

  Widget _buildTransferForm() {
    return Form(
      key: _key,
      child: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: <Widget>[
            DropdownButton<String>(
              value: selectedBankCode,
              hint: const Text('Select a bank'),
              onChanged: (String? newValue) {
                setState(() {
                  selectedBankCode = newValue;
                });
              },
              items: bankList.map<DropdownMenuItem<String>>((Bank bank) {
                return DropdownMenuItem<String>(
                  value: bank.bankCode.toString(),
                  child: Text(bank.bankName.toString()),
                );
              }).toList(),
            ),
            const SizedBox(height: 6),
            Text(
              'Selected Bank Code: $selectedBankCode',
              style: const TextStyle(color: Colors.green, fontSize: 15),
            ),
            const SizedBox(height: 6),
            TextFormField(
              decoration: const InputDecoration(
                labelText: 'Account No.',
                filled: true,
                isDense: true,
              ),
              controller: _accountController,
              keyboardType: TextInputType.number,
              maxLength: 10,
              onChanged: (value) {
                if (value.length == 10) {
                  fetchAccountName(value);
                } else {
                  setState(() {
                    accountName = null;
                  });
                }
              },
              autocorrect: false,
            ),
            const SizedBox(
              height: 4,
            ),
            Text(
              'Account Name: ${accountName ?? 'N/A'}',
              style: const TextStyle(color: Colors.green, fontSize: 15),
            ),
            const SizedBox(
              height: 6,
            ),
            Text(
              'Session ID: ${session_id ?? 'N/A'}',
              style: const TextStyle(color: Colors.green, fontSize: 15),
            ),
            Text(
              'New Bank Code: ${bank_codes ?? 'N/A'}',
              style: const TextStyle(color: Colors.green, fontSize: 15),
            ),
            Text(
              'Response Code: ${response_codes ?? 'N/A'}',
              style: const TextStyle(color: Colors.green, fontSize: 15),
            ),
            const SizedBox(
              height: 6,
            ),
            TextFormField(
              decoration: const InputDecoration(
                labelText: 'Amount',
                filled: true,
                isDense: true,
              ),
              keyboardType: TextInputType.number,
              controller: _amountController,
              //validator: (val) => _validateRequired(val, 'Password'),
            ),
            const SizedBox(
              height: 16,
            ),
            TextFormField(
              decoration: const InputDecoration(
                labelText: 'Narration',
                filled: true,
                isDense: true,
              ),
              controller: _narrationController,
              //validator: (val) => _validateRequired(val, 'Password'),
            ),
            const SizedBox(
              height: 16,
            ),
            MaterialButton(
              color: Theme.of(context).primaryColor,
              textColor: Colors.white,
              padding: const EdgeInsets.all(16),
              shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(8.0)),
              child: const Text('Make Transfer'),
              onPressed: () {
                performMoneyTransfer(context);
              },
            ),
          ],
        ),
      ),
    );
  }

  Future<void> fetchAccountName(String accountNumber) async {
    try {
      String requestRef = Random().nextInt(100000).toString();
      Map<String, dynamic> data = {
        'beneficiaryAccountNumber': accountNumber,
        'beneficiaryBankCode': selectedBankCode,
        'SenderTrackingReference': '',
        'isRequestFromVirtualAccount': true,
      };
      NameInquiryResponse response =
          await KudaBank().name_inquiry(data, requestRef);

      print("Request Reference " + requestRef);

      setState(() {
        accountName = response.nameInquiry!.beneficiaryName;
        session_id = response.nameInquiry!.sessionID;
        bank_codes = response.nameInquiry!.beneficiaryBankCode;
        response_codes = response.message;
      });
    } catch (error) {
      print('Error fetching account name: $error');
    }
  }

  Future<void> performMoneyTransfer(BuildContext context) async {
    try {
      String accountNumber = _accountController.text;
      String? bankCode = selectedBankCode;
      String? accountNames = accountName;
      String? sessionID = session_id;
      int amount = int.parse(_amountController.text);
      String narration = _narrationController.text;

      Map<String, dynamic> data = {
        'ClientAccountNumber': '3000002674',
        'beneficiarybankCode': bankCode,
        'beneficiaryAccount': accountNumber,
        'beneficiaryName': accountNames,
        'amount': amount * 100,
        'narration': narration,
        'nameEnquirySessionID': sessionID,
        'trackingReference': '',
        'senderName': accountNames,
        'clientFeeCharge': 0,
      };

      print("Request Reference " + requestRef);

      SendMoneyKuda response =
          await KudaBank().single_fund_transfer(data, requestRef);

      // Check the response for success
      if (response.status == true) {
        showDialog(
          context: context,
          builder: (_) => AlertDialog(
            title: const Text('Money Transfer'),
            content: const Text('Transfer successful!'),
            actions: [
              ElevatedButton(
                onPressed: () {
                  Navigator.pop(context);
                },
                child: const Text('OK'),
              ),
            ],
          ),
        );
      } else {
        showDialog(
          context: context,
          builder: (_) => AlertDialog(
            title: const Text('Money Transfer'),
            content: const Text('Transfer failed!'),
            actions: [
              ElevatedButton(
                onPressed: () {
                  Navigator.pop(context);
                },
                child: const Text('OK'),
              ),
            ],
          ),
        );
      }
    } catch (error) {
      showDialog(
        context: context,
        builder: (_) => AlertDialog(
          title: const Text('Money Transfer'),
          content: Text('Transfer failed: ${error.toString()}'),
          actions: [
            ElevatedButton(
              onPressed: () {
                Navigator.pop(context);
              },
              child: const Text('OK'),
            ),
          ],
        ),
      );
    }
  }
}
